Handle existing privileges for grant and revoke

amrutagokhale opened the pull request:
At present, the grant and revoke code fails when trying to grant a privilege that the user already possesses or revoking a privilege that a user doesn't possess. This PR adds code for handing these cases. Also added a commit for cleaning up the existing product test. The two commits are in the following order:
1. Cleaning up the product test for grant and revoke
2. Handling existing privileges in grant and revoke statements.

Testing:
Modified the existing product test for grant/revoke.

TD internal review is here: https://github.com/Teradata/presto/pull/340

CC: @dain

martint commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -564,9 +566,21 @@ public void alterPartition(String databaseName, String tableName, Partition part
+        checkArgument(!containsAllPrivilege(requestedPrivileges), "\"ALL\" not supported in PrivilegeGrantInfo.privilege");
+
         try {
+            Set<HivePrivilegeInfo> existingPrivileges = getTablePrivileges(grantee, databaseName, tableName);
```
Equality/hashCode for HivePrivilegeInfo is based on the privilege type + grant option. What's the expected behavior if the table already has, say, "SELECT with grant option" and the call to this method requests "SELECT without grant option"?

amrutagokhale commented:
@martint I updated the code to handle those cases. Can you please take a look?

This is how `GRANT` behaves now:
1) If the user possesses `SELECT with grant option`, and there is a subsequent call to  grant `SELECT without grant option`, the command returns without any error (since the user already has the requested privilege).
2) If the users possesses `SELECT without grant option` and there is a subsequent call to grant `SELECT with grant option`, the command throws an exception (since the underlying Hive metastore client API doesn't support this case and will eventually result in an error).

The case of `REVOKE` is different. I verified that the Hive metastore client API doesn't throw an error when revoking the requested privilege, as long as there is an existing `HivePrivilegeInfo` whose `HivePrivilege` part matches with that of the requested `HivePrivilegeInfo` (irrespective of the `GRANT OPTION` flag in the two privileges). In effect, it revokes the existing privilege.

So I changed the `REVOKE` code so as to keep the requested privileges in the `requestedPrivileges` set, as long as the above condition is met. 

I have also modified the product test to test the new behavior by adding `WITH GRANT OPTION` in `GRANT`.

martint commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/HivePrivilegeInfo.java:
```diff
@@ -96,6 +96,18 @@ public static HivePrivilege toHivePrivilege(Privilege privilege)
 
+    public boolean isContainedIn(HivePrivilegeInfo hivePrivilegeInfo)
+    {
+        if (this.getHivePrivilege().equals(hivePrivilegeInfo.getHivePrivilege()) &&
```
You can just do:

```java
return (this.getHivePrivilege.equals(....) && ...)
```

amrutagokhale replied:
done

martint commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -564,9 +569,34 @@ public void alterPartition(String databaseName, String tableName, Partition part
+
+                for (HivePrivilegeInfo existingPrivilege : existingPrivileges) {
+                    if ((requestedPrivilege.isContainedIn(existingPrivilege))) {
+                        iterator.remove();
```
It's not a good idea to modify the set that's passed in:
1. This could fail if the set is immutable
2. If the caller continues to use the set after the call, it will cause subtle corruption bugs.

amrutagokhale commented:
@martint Addressed comments.

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/HivePrivilegeInfo.java:
```diff
@@ -96,6 +96,13 @@ public static HivePrivilege toHivePrivilege(Privilege privilege)
 
+    public boolean isContainedIn(HivePrivilegeInfo hivePrivilegeInfo)
+    {
+        return (this.getHivePrivilege().equals(hivePrivilegeInfo.getHivePrivilege()) &&
```
The "this." are not needed and should be removed

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -604,9 +635,24 @@ public void grantTablePrivileges(String databaseName, String tableName, String g
+
         try {
+            Set<HivePrivilege> existingHivePrivileges = getTablePrivileges(grantee, databaseName, tableName)
+                    .stream()
```
Put .stream() on previous line

amrutagokhale commented:
@electrum Addressed your comments.

martint commented:
Merged, thanks!
